akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGroupBySpec.scala

```diff
@@ -552,24 +553,19 @@ class FlowGroupBySpec extends StreamSpec("""
     }
 
     "work with random demand" in assertAllStagesStopped {
-      var blockingNextElement: ByteString = null.asInstanceOf[ByteString]
-
-      val probes = new java.util.ArrayList[Promise[TestSubscriber.Probe[ByteString]]](100)
-      (0 to 99).foreach(_ => probes.add(Promise[TestSubscriber.Probe[ByteString]]()))
-
-      var probesWriterTop = 0
-      var probesReaderTop = 0
-
-      case class SubFlowState(probe: TestSubscriber.Probe[ByteString], hasDemand: Boolean, firstElement: ByteString)
-      val map = new util.HashMap[Int, SubFlowState]()
+      val probes = IndexedSeq.tabulate(100)(_ => Promise[TestSubscriber.Probe[ByteString]]())
 
       final class ProbeSink(val attributes: Attributes, shape: SinkShape[ByteString])(implicit system: ActorSystem)
           extends SinkModule[ByteString, TestSubscriber.Probe[ByteString]](shape) {
+
+        // materialized on demand by GroupBy so we need thread safety here
+        val materializationCounter = new AtomicInteger(0)
```

Comment:
The thread unsafe counter in `probesTopWrite` (and array population) was not causing the issue but still worth fixing